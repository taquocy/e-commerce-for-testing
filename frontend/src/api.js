import axios from "axios";

axios.interceptors.request.use(
  function (config) {
    const { origin } = new URL(config.url);

    const allowedOrigins = [process.env.REACT_APP_BASE_ENDPOINT];
    const token = localStorage.getItem("access-token");

    if (allowedOrigins.includes(origin)) {
      config.headers.authorization = token;
    }
    return config;
  },
  function (error) {
    return Promise.reject(error);
  }
);

// Fetch Product List
export const fetchProductList = async ({ pageParam = 1 }) => {
  const { data } = await axios.get(
    `${process.env.REACT_APP_BASE_ENDPOINT}/product?page=${pageParam}`
  );

  return data;
};

// Fetch Product by ID
export const fetchProduct = async (id) => {
  const { data } = await axios.get(
    `${process.env.REACT_APP_BASE_ENDPOINT}/product/${id}`
  );

  return data;
};

// Create Product
export const postProduct = async (input) => {
  const { data } = await axios.post(
    `${process.env.REACT_APP_BASE_ENDPOINT}/product/`,
    input
  );

  return data;
};

// Register User
export const fetcRegister = async (input) => {
  const { data } = await axios.post(
    `${process.env.REACT_APP_BASE_ENDPOINT}/auth/register`,
    input
  );

  return data;
};

// Login User
export const fetchLogin = async (input) => {
  const { data } = await axios.post(
    `${process.env.REACT_APP_BASE_ENDPOINT}/auth/login`,
    input
  );

  return data;
};

// Fetch Current User
export const fetchMe = async () => {
  const { data } = await axios.get(
    `${process.env.REACT_APP_BASE_ENDPOINT}/auth/me`
  );
  return data;
};

// Logout User
export const fetchLogout = async () => {
  const { data } = await axios.post(
    `${process.env.REACT_APP_BASE_ENDPOINT}/auth/logout`,
    {
      refresh_token: localStorage.getItem("refresh-token"),
    }
  );
  return data;
};

// Post Order
export const postOrder = async (input) => {
  const { data } = await axios.post(
    `${process.env.REACT_APP_BASE_ENDPOINT}/order`,
    input
  );
  return data;
};

// Fetch Orders
export const fetchOrders = async () => {
  const { data } = await axios.get(
    `${process.env.REACT_APP_BASE_ENDPOINT}/order`
  );
  return data;
};

// Delete Product
export const deleteProduct = async (product_id) => {
  const { data } = await axios.delete(
    `${process.env.REACT_APP_BASE_ENDPOINT}/product/${product_id}`
  );

  return data;
};

// Update Product
export const updateProduct = async (input, product_id) => {
  const { data } = await axios.put(
    `${process.env.REACT_APP_BASE_ENDPOINT}/product/${product_id}`,
    input
  );

  return data;
};

// Forgot Password API
export const fetchForgotPassword = async (email) => {
  console.log("Calling API:", `${process.env.REACT_APP_BASE_URL}/auth/forgot-password`);
  const { data } = await axios.post(`${process.env.REACT_APP_BASE_URL}/auth/forgot-password`, { email });
  return data;
};

// Reset Password API
export const fetchResetPassword = async (token, newPassword) => {
  const { data } = await axios.post(
    `${process.env.REACT_APP_BASE_ENDPOINT}/auth/reset-password`,
    {
      token,
      newPassword,
    }
  );

  return data;
};
