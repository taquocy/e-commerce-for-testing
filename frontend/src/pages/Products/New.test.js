import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { QueryClient, QueryClientProvider } from "react-query";
import { BrowserRouter } from "react-router-dom";
import NewProduct from "./New";
import { postProduct } from "../../api";
import { message } from "antd";

// Mock cÃ¡c dependencies
jest.mock("../../api");
jest.mock("antd", () => ({
  message: {
    loading: jest.fn(),
    success: jest.fn(),
    error: jest.fn(),
  },
}));

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const wrapper = ({ children }) => (
  <QueryClientProvider client={queryClient}>
    <BrowserRouter>{children}</BrowserRouter>
  </QueryClientProvider>
);

describe("NewProduct Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  // Test 8: Validation for negative price
  test("shows validation error for negative price", async () => {
    render(<NewProduct />, { wrapper });

    await userEvent.type(screen.getByLabelText(/Price/i), "-10");

    fireEvent.click(screen.getByText("Add Product"));

    await waitFor(() => {
      expect(screen.getByText("Price must be greater than 0")).toBeInTheDocument();
    });
  });

  // Test 9: Handles API error gracefully
  test("displays error message on API failure", async () => {
    postProduct.mockRejectedValueOnce(new Error("Failed to add product"));

    const { container } = render(<NewProduct />, { wrapper });

    await userEvent.type(screen.getByLabelText(/Title/i), "Test Product");
    await userEvent.type(screen.getByLabelText(/Description/i), "Test Description");
    await userEvent.type(screen.getByLabelText(/Price/i), "99.99");

    fireEvent.click(screen.getByText("Add Product"));

    await waitFor(() => {
      expect(message.loading).toHaveBeenCalledWith({
        content: "Loading...",
        key: "product_update",
      });

      expect(message.error).toHaveBeenCalledWith({
        content: "Failed to add product",
        key: "product_update",
        duration: 2,
      });
    });
  });

  // Test 10: Redirects to admin page on successful submission
  test("redirects to admin page after successful submission", async () => {
    postProduct.mockResolvedValueOnce({ success: true });

    const mockNavigate = jest.fn();
    jest.mock("react-router-dom", () => ({
      ...jest.requireActual("react-router-dom"),
      useNavigate: () => mockNavigate,
    }));

    const { container } = render(<NewProduct />, { wrapper });

    await userEvent.type(screen.getByLabelText(/Title/i), "Test Product");
    await userEvent.type(screen.getByLabelText(/Description/i), "Test Description");
    await userEvent.type(screen.getByLabelText(/Price/i), "99.99");

    fireEvent.click(screen.getByText("Add Product"));

    await waitFor(() => {
      expect(postProduct).toHaveBeenCalledWith({
        title: "Test Product",
        description: "Test Description",
        price: "99.99",
        photos: "[]",
      });

      expect(mockNavigate).toHaveBeenCalledWith("/admin");
    });
  });
});
