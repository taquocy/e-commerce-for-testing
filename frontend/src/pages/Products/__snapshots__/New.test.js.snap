// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewProduct Component renders NewProduct component 1`] = `
<DocumentFragment>
  <div>
    <nav>
      <ul
        class="admin-menu"
      >
        <li>
          <a
            href="/admin"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/admin/orders"
          >
            Order
          </a>
        </li>
        <li>
          <a
            href="/admin/products"
          >
            Products
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="css-1q7njkh"
    >
      <button
        class="chakra-button css-eaw8jh"
        type="button"
      >
        Edit
      </button>
      <div
        class="css-0"
      >
        <div
          class="css-lgjuw7"
        >
          <form>
            <div
              class="chakra-form-control css-0"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r0:"
                id="field-:r0:-label"
              >
                Title
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required="true"
                class="chakra-input css-0"
                id="field-:r0:"
                name="title"
                required=""
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r1:"
                id="field-:r1:-label"
              >
                Description
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <textarea
                aria-required="true"
                class="chakra-textarea css-0"
                id="field-:r1:"
                name="description"
                required=""
              />
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r2:"
                id="field-:r2:-label"
              >
                Price
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required="true"
                class="chakra-input css-0"
                id="field-:r2:"
                name="price"
                required=""
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r3:"
                id="field-:r3:-label"
              >
                Currency
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  aria-required="true"
                  class="chakra-select css-1ik61og"
                  id="field-:r3:"
                  name="currency"
                  required=""
                >
                  <option
                    value="USD"
                  >
                    USD
                  </option>
                  <option
                    value="EUR"
                  >
                    EUR
                  </option>
                  <option
                    value="VND"
                  >
                    VND
                  </option>
                  <option
                    value="GBP"
                  >
                    GBP
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r4:"
                id="field-:r4:-label"
              >
                Photos
              </label>
              <div>
                <button
                  class="chakra-button css-10tmkll"
                  type="button"
                >
                  Add a Photo
                </button>
              </div>
            </div>
            <button
              class="chakra-button css-2cxz8l"
              type="submit"
            >
              Add Product
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`NewProduct Component renders navigation links 1`] = `
<DocumentFragment>
  <div>
    <nav>
      <ul
        class="admin-menu"
      >
        <li>
          <a
            href="/admin"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/admin/orders"
          >
            Order
          </a>
        </li>
        <li>
          <a
            href="/admin/products"
          >
            Products
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="css-1q7njkh"
    >
      <button
        class="chakra-button css-eaw8jh"
        type="button"
      >
        Edit
      </button>
      <div
        class="css-0"
      >
        <div
          class="css-lgjuw7"
        >
          <form>
            <div
              class="chakra-form-control css-0"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r5:"
                id="field-:r5:-label"
              >
                Title
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required="true"
                class="chakra-input css-0"
                id="field-:r5:"
                name="title"
                required=""
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r6:"
                id="field-:r6:-label"
              >
                Description
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <textarea
                aria-required="true"
                class="chakra-textarea css-0"
                id="field-:r6:"
                name="description"
                required=""
              />
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r7:"
                id="field-:r7:-label"
              >
                Price
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required="true"
                class="chakra-input css-0"
                id="field-:r7:"
                name="price"
                required=""
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r8:"
                id="field-:r8:-label"
              >
                Currency
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  aria-required="true"
                  class="chakra-select css-1ik61og"
                  id="field-:r8:"
                  name="currency"
                  required=""
                >
                  <option
                    value="USD"
                  >
                    USD
                  </option>
                  <option
                    value="EUR"
                  >
                    EUR
                  </option>
                  <option
                    value="VND"
                  >
                    VND
                  </option>
                  <option
                    value="GBP"
                  >
                    GBP
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="chakra-form-control css-1fhgjcy"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="field-:r9:"
                id="field-:r9:-label"
              >
                Photos
              </label>
              <div>
                <button
                  class="chakra-button css-10tmkll"
                  type="button"
                >
                  Add a Photo
                </button>
              </div>
            </div>
            <button
              class="chakra-button css-2cxz8l"
              type="submit"
            >
              Add Product
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
